pkg/ingester/ingester_forced_compaction_test.go

```diff
@@ -0,0 +1,573 @@
+			require.NoError(t, err)
+			require.NoError(t, services.StartAndAwaitRunning(ctx, ingester))
+			t.Cleanup(func() {
+				require.NoError(t, services.StopAndAwaitTerminated(ctx, ingester))
+			})
+
+			// Wait until it's ACTIVE.
+			test.Poll(t, time.Second, ring.ACTIVE, func() interface{} {
+				return ingester.lifecycler.GetState()
+			})
+
+			// Keep track of the last timestamp written by each writer.
+			writerTimesMx := sync.Mutex{}
+			writerTimes := make([]int64, numWriters)
+
+			// Start writers.
+			wg.Add(numWriters)
+			for i := 0; i < numWriters; i++ {
+				go func(writerID int) {
+					defer wg.Done()
+
+					// Decide the series written by this writer.
+					fromSeriesID := writerID * (numSeries / numWriters)
+					toSeriesID := fromSeriesID + (numSeries / numWriters) - 1
+
+					for sampleIdx := 0; sampleIdx < numSamplesPerSeries; sampleIdx++ {
+						timestamp := startTime.Add(time.Duration(sampleIdx) * time.Millisecond).UnixMilli()
+
+						// Prepare the series to write.
+						seriesToWrite := make([]series, 0, toSeriesID-fromSeriesID+1)
+						for seriesIdx := fromSeriesID; seriesIdx <= toSeriesID; seriesIdx++ {
+							seriesToWrite = append(seriesToWrite, series{
+								lbls:      labels.FromStrings(labels.MetricName, fmt.Sprintf("series_%05d", seriesIdx)),
+								value:     float64(sampleIdx),
+								timestamp: timestamp,
+							})
+						}
+
+						require.NoError(t, pushSeriesToIngester(ctxWithUser, t, ingester, seriesToWrite))
+
+						// Keep track of the last timestamp written.
+						writerTimesMx.Lock()
+						writerTimes[writerID] = timestamp
+						writerTimesMx.Unlock()
+					}
+				}(i)
+			}
+
+			// Start readers (each reader reads all series).
+			wg.Add(numReaders)
+			for i := 0; i < numReaders; i++ {
+				go func() {
+					defer wg.Done()
+
+					s := stream{ctx: ctxWithUser}
+					err := ingester.QueryStream(readerReq, &s)
+					require.NoError(t, err)
+
+					res, err := client.StreamsToMatrix(model.Earliest, model.Latest, s.responses)
+					require.NoError(t, err)
+
+					// We expect a response consistent with the samples written.
+					for entryIdx, entry := range res {
+						for sampleIdx, sample := range entry.Values {
+							expectedTime := model.Time(startTime.Add(time.Duration(sampleIdx) * time.Millisecond).UnixMilli())
+							expectedValue := model.SampleValue(sampleIdx)
+
+							require.Equalf(t, expectedTime, sample.Timestamp, "response entry: %d series: %s sample idx: %d", entryIdx, entry.Metric.String(), sampleIdx)
+							require.Equalf(t, expectedValue, sample.Value, "response entry: %d series: %s sample idx: %d", entryIdx, entry.Metric.String(), sampleIdx)
+						}
+					}
+				}()
+			}
+
+			// Start a goroutine continuously forcing the TSDB head compaction.
+			stopForcedCompaction := make(chan struct{})
+			go func() {
+				for {
+					select {
+					case <-stopForcedCompaction:
+						return
+					case <-time.After(100 * time.Millisecond):
```

Comment:
This confuses me a bit. If I understand correctly, if the readers and writers all complete within 100ms, this will never be run. Is that right?

Reply:
That's right, but then later I do check "Ensure at least 1 forced compaction has been done".